src/wasi.rs

```diff
@@ -0,0 +1,792 @@
+pub const __WASI_SIGABRT: u8 = 6;
+pub const __WASI_SIGBUS: u8 = 7;
+pub const __WASI_SIGFPE: u8 = 8;
+pub const __WASI_SIGKILL: u8 = 9;
+pub const __WASI_SIGUSR1: u8 = 10;
+pub const __WASI_SIGSEGV: u8 = 11;
+pub const __WASI_SIGUSR2: u8 = 12;
+pub const __WASI_SIGPIPE: u8 = 13;
+pub const __WASI_SIGALRM: u8 = 14;
+pub const __WASI_SIGTERM: u8 = 15;
+pub const __WASI_SIGCHLD: u8 = 16;
+pub const __WASI_SIGCONT: u8 = 17;
+pub const __WASI_SIGSTOP: u8 = 18;
+pub const __WASI_SIGTSTP: u8 = 19;
+pub const __WASI_SIGTTIN: u8 = 20;
+pub const __WASI_SIGTTOU: u8 = 21;
+pub const __WASI_SIGURG: u8 = 22;
+pub const __WASI_SIGXCPU: u8 = 23;
+pub const __WASI_SIGXFSZ: u8 = 24;
+pub const __WASI_SIGVTALRM: u8 = 25;
+pub const __WASI_SIGPROF: u8 = 26;
+pub const __WASI_SIGWINCH: u8 = 27;
+pub const __WASI_SIGPOLL: u8 = 28;
+pub const __WASI_SIGPWR: u8 = 29;
+pub const __WASI_SIGSYS: u8 = 30;
+pub const __WASI_SUBSCRIPTION_CLOCK_ABSTIME: u16 = 0x0001;
+pub const __WASI_WHENCE_CUR: u8 = 0;
+pub const __WASI_WHENCE_END: u8 = 1;
+pub const __WASI_WHENCE_SET: u8 = 2;
+
+#[cfg_attr(feature = "extra_traits", derive(Debug))]
+pub enum FILE {}
+impl ::Copy for FILE {}
+impl ::Clone for FILE {
+    fn clone(&self) -> FILE {
+        *self
+    }
+}
+
+s! {
+    #[repr(align(8))]
+    pub struct fpos_t {
+        data: [u8; 16],
+    }
+
+    pub struct tm {
+        pub tm_sec: c_int,
+        pub tm_min: c_int,
+        pub tm_hour: c_int,
+        pub tm_mday: c_int,
+        pub tm_mon: c_int,
+        pub tm_year: c_int,
+        pub tm_wday: c_int,
+        pub tm_yday: c_int,
+        pub tm_isdst: c_int,
+        pub __tm_gmtoff: c_int,
+        pub __tm_zone: *const c_char,
+        pub __tm_nsec: c_int,
+    }
+
+    pub struct timespec {
+        pub tv_sec: time_t,
+        pub tv_nsec: c_long,
+    }
+
+    pub struct itimerspec {
+        pub it_interval: timespec,
+        pub it_value: timespec,
+    }
+
+    pub struct __clockid {}
+
+    pub struct __wasi_dirent_t {
+        pub d_next: __wasi_dircookie_t,
+        pub d_ino: __wasi_inode_t,
+        pub d_namlen: u32,
+        pub d_type: __wasi_filetype_t,
+    }
+
+    pub struct __wasi_event_u_fd_readwrite_t {
+        pub nbytes: __wasi_filesize_t,
+        pub flags: __wasi_eventrwflags_t,
+    }
+
+    pub struct __wasi_fdstat_t {
+        pub fs_filetype: __wasi_filetype_t,
+        pub fs_flags: __wasi_fdflags_t,
+        pub fs_rights_base: __wasi_rights_t,
+        pub fs_rights_inheriting: __wasi_rights_t,
+    }
+
+    pub struct __wasi_filestat_t {
+        pub st_dev: __wasi_device_t,
+        pub st_ino: __wasi_inode_t,
+        pub st_filetype: __wasi_filetype_t,
+        pub st_nlink: __wasi_linkcount_t,
+        pub st_size: __wasi_filesize_t,
+        pub st_atim: __wasi_timestamp_t,
+        pub st_mtim: __wasi_timestamp_t,
+        pub st_ctim: __wasi_timestamp_t,
+    }
+
+    pub struct __wasi_ciovec_t {
+        pub buf: *const ::c_void,
+        pub buf_len: usize,
```

Comment:
I'd prefer if these would use the "right" C types (e.g. `size_t`, `ptrdiff_t`, `uintptr_t`, etc.) Same pretty much everywhere where `usize` is used below.

Reply:
This is actually all generated by `bindgen` where `size_t` is the same as `usize`. I think they're basically interchangeable at this point (wasi is unlikely to ever have the insanity that is differences in the three types you mentioned), but would you still prefer `s/usize/size_t/` everywhere?

Reply:
They are definitely interchangeable for this target, I think that's all that matters. My preference for `size_t` here is for clarity, since that's what the C code does. IIRC bindgen typically preserves the C types if one explicitly tells it to use the `libc::` types, but... there is no libc for wasi yet... If bindgen also generated the `size_t` type alias, it maybe has a "bug".

Reply:
Ok, I'll go ahead and switch it all to `size_t`